### ISSUE TITLE
citar-open fails when invoked on citation via embark-act

### ISSUE THREAD

**Issue by jkroes** [bdarcus/citar#584]
**Describe the bug**
Calling embark-act and selecting citar-run-default-action (with citar-default-action equal to citar-open) on a citation fails, because the key is a string, not a list. 

**Backtrace**
```
Debugger entered--Lisp error: (wrong-type-argument stringp 80)
  string-match(" & " 80 0)
  split-string(80 " & ")
  (seq-map (lambda (key) (cdr (seq-find (lambda (cand-key-entry) (string= key (cadr cand-key-entry))) candidates))) (split-string key-entry " & "))
  (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find (lambda (cand-key-entry) (string= key (cadr cand-key-entry))) candidates))) (split-string key-entry " & ")))
  (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find (lambda (cand-key-entry) (string= key ...)) candidates))) (split-string key-entry " & "))))
  (lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find ... candidates))) (split-string key-entry " & ")))))(80)
  mapcar((lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find ... candidates))) (split-string key-entry " & "))))) "PauseNotEnd2013")
  #f(compiled-function #'sequence #<bytecode 0x4d4978e7>)((lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find ... candidates))) (split-string key-entry " & "))))) "PauseNotEnd2013")
  apply(#f(compiled-function #'sequence #<bytecode 0x4d4978e7>) (lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find ... candidates))) (split-string key-entry " & "))))) "PauseNotEnd2013" nil)
  seq-map((lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find ... candidates))) (split-string key-entry " & "))))) "PauseNotEnd2013")
  seq-mapcat((lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr (seq-find ... candidates))) (split-string key-entry " & "))))) "PauseNotEnd2013")
  (let ((candidates (citar--get-candidates))) (seq-mapcat (lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda (key) (cdr ...)) (split-string key-entry " & "))))) keys-entries))
  (if (seq-every-p #'consp keys-entries) keys-entries (let ((candidates (citar--get-candidates))) (seq-mapcat (lambda (key-entry) (if (consp key-entry) (list key-entry) (seq-remove #'null (seq-map (lambda ... ...) (split-string key-entry " & "))))) keys-entries)))
  citar--ensure-entries("PauseNotEnd2013")
  (let* ((embark-default-action-overrides '((multi-category . citar-open-multi) (file . citar-file-open) (url . browse-url))) (key-entry-alist (citar--ensure-entries keys-entries)) (files (citar-file--files-for-multiple-entries key-entry-alist (append citar-library-paths citar-notes-paths) nil)) (links (seq-map (lambda (key-entry) (citar-get-link (cdr key-entry))) key-entry-alist)) (resource-candidates (delete-dups (append files (remq nil links))))) (cond ((eq nil resource-candidates) (error "No associated resources")) ((unless citar-open-prompt (eq 1 (length resource-candidates))) (citar-open-multi (car resource-candidates))) (t (citar-open-multi (citar-select-resource files links)))))
  citar-open("PauseNotEnd2013")
  funcall(citar-open "PauseNotEnd2013")
  citar-run-default-action("PauseNotEnd2013")
  #f(compiled-function () #<bytecode 0x45d50aef>)()
  apply(#f(compiled-function () #<bytecode 0x45d50aef>) nil)
  embark--quit-and-run(#f(compiled-function () #<bytecode 0x45d50aef>))
  embark--act(citar-run-default-action (:orig-type citar-citation :orig-target "PauseNotEnd2013" :bounds (103 . 126) :type citar-citation :target "PauseNotEnd2013") t)
  (condition-case err (embark--act action (if (and (eq action default-action) (eq action embark--command)) (plist-put (plist-put (copy-sequence target) :target (plist-get target :orig-target)) :type (plist-get target :orig-type)) target) (if embark-quit-after-action (not arg) arg)) (user-error (funcall (if repeat #'message #'user-error) "%s" (cadr err))))
  (let ((repeat (memq action embark-repeat-actions))) (unless repeat (mapc #'funcall indicators)) (condition-case err (embark--act action (if (and (eq action default-action) (eq action embark--command)) (plist-put (plist-put (copy-sequence target) :target (plist-get target :orig-target)) :type (plist-get target :orig-type)) target) (if embark-quit-after-action (not arg) arg)) (user-error (funcall (if repeat #'message #'user-error) "%s" (cadr err)))) (when-let (new-targets (and repeat (embark--targets))) (setq default-done #'embark-done targets (embark--rotate new-targets (or (cl-position-if (let (...) (lambda ... ...)) new-targets) 0)))))
  (cond ((eq action #'embark-act) (message "Press an action key")) ((eq action #'embark-cycle) (setq targets (embark--rotate targets (prefix-numeric-value prefix-arg)))) (t (let ((repeat (memq action embark-repeat-actions))) (unless repeat (mapc #'funcall indicators)) (condition-case err (embark--act action (if (and (eq action default-action) (eq action embark--command)) (plist-put (plist-put ... :target ...) :type (plist-get target :orig-type)) target) (if embark-quit-after-action (not arg) arg)) (user-error (funcall (if repeat #'message #'user-error) "%s" (cadr err)))) (when-let (new-targets (and repeat (embark--targets))) (setq default-done #'embark-done targets (embark--rotate new-targets (or (cl-position-if ... new-targets) 0)))))))
  (let* ((target (car targets)) (action (or (embark--prompt indicators (let ((embark-default-action-overrides ...)) (embark--action-keymap (plist-get target :type) (cdr targets))) targets) (user-error "Canceled"))) (default-action (or default-done (embark--default-action (plist-get target :type))))) (cond ((eq action #'embark-act) (message "Press an action key")) ((eq action #'embark-cycle) (setq targets (embark--rotate targets (prefix-numeric-value prefix-arg)))) (t (let ((repeat (memq action embark-repeat-actions))) (unless repeat (mapc #'funcall indicators)) (condition-case err (embark--act action (if (and ... ...) (plist-put ... :type ...) target) (if embark-quit-after-action (not arg) arg)) (user-error (funcall (if repeat ... ...) "%s" (cadr err)))) (when-let (new-targets (and repeat (embark--targets))) (setq default-done #'embark-done targets (embark--rotate new-targets (or ... 0))))))))
  (while (let* ((target (car targets)) (action (or (embark--prompt indicators (let (...) (embark--action-keymap ... ...)) targets) (user-error "Canceled"))) (default-action (or default-done (embark--default-action (plist-get target :type))))) (cond ((eq action #'embark-act) (message "Press an action key")) ((eq action #'embark-cycle) (setq targets (embark--rotate targets (prefix-numeric-value prefix-arg)))) (t (let ((repeat (memq action embark-repeat-actions))) (unless repeat (mapc #'funcall indicators)) (condition-case err (embark--act action (if ... ... target) (if embark-quit-after-action ... arg)) (user-error (funcall ... "%s" ...))) (when-let (new-targets (and repeat ...)) (setq default-done #'embark-done targets (embark--rotate new-targets ...))))))))
  (unwind-protect (while (let* ((target (car targets)) (action (or (embark--prompt indicators (let ... ...) targets) (user-error "Canceled"))) (default-action (or default-done (embark--default-action (plist-get target :type))))) (cond ((eq action #'embark-act) (message "Press an action key")) ((eq action #'embark-cycle) (setq targets (embark--rotate targets (prefix-numeric-value prefix-arg)))) (t (let ((repeat ...)) (unless repeat (mapc ... indicators)) (condition-case err (embark--act action ... ...) (user-error ...)) (when-let (new-targets ...) (setq default-done ... targets ...))))))) (mapc #'funcall indicators))
  (let* ((targets (or (embark--targets) (user-error "No target found"))) (indicators (mapcar #'funcall embark-indicators)) (default-done nil)) (when (and arg (not (minibufferp))) (setq targets (embark--rotate targets (prefix-numeric-value arg)))) (unwind-protect (while (let* ((target (car targets)) (action (or (embark--prompt indicators ... targets) (user-error "Canceled"))) (default-action (or default-done (embark--default-action ...)))) (cond ((eq action #'embark-act) (message "Press an action key")) ((eq action #'embark-cycle) (setq targets (embark--rotate targets ...))) (t (let (...) (unless repeat ...) (condition-case err ... ...) (when-let ... ...)))))) (mapc #'funcall indicators)))
  embark-act(nil)
  #<subr funcall-interactively>(embark-act nil)
  apply(#<subr funcall-interactively> (embark-act nil))
  funcall-interactively(embark-act nil)
  #<subr call-interactively>(embark-act nil nil)
  apply(#<subr call-interactively> (embark-act nil nil))
  explain-pause--wrap-call-interactively(#<subr call-interactively> embark-act nil nil)
  apply(explain-pause--wrap-call-interactively #<subr call-interactively> (embark-act nil nil))
  call-interactively(embark-act nil nil)
  command-execute(embark-act)
```

**To Reproduce**
1. Go to a citation.
2. Call embark-act.
3. Select citar-run-default-action.

**Expected behavior**
The default action, citar-open, executes with the citation at point. 

**Emacs version:**
GNU Emacs 27.2 (build 1, aarch64-apple-darwin21.1.0, NS appkit-2113.00 Version 12.0.1 (Build 21A559)) of 2021-12-26

**Comment by jkroes** [comment]
I see this in org-mode.  embark-dwim explicitly wraps the citation key in a list before passing to citar-run-default-action, but embark-act does not.

**Comment by bdarcus** [comment]
I'm admittedly a bit tired, so may just be missing something obvious, but ...

So is the issue `embark-act`, or is it `citar-run-default-action`?

I don't have `citar-open` as my default action. If I change it to that, and hit RET after `embark-act`, I do get odd behavior, but not any error or backtrace as you do.

And then there's this, which I can't understand (this resource does have an associated file, which is presented when running `citar-open` directly, but not, as you can see, `citar-run-default-action`).

```elisp
ELISP> (citar--get-entry "stanley2021")
(("file" . "/home/bruce/Zotero/storage/7HH79WKH/2021 - Movie at the Ellipse A Study in Fascist Propagand.html")
 ("keywords" . "2021-01-06,far right,fascism,insurrection,media,riots,Trump,United States,Washington (DC)")
 ("url" . "https://www.justsecurity.org/74504/movie-at-the-ellipse-a-study-in-fascist-propaganda/")
 ("date" . "2021-02-04T13:30:28+00:00")
 ("author" . "Stanley, Jason")
 ("title" . "Movie at the {{Ellipse}}: A Study in Fascist Propaganda")
 ("=type=" . "online")
 ("=key=" . "stanley2021"))

ELISP> (citar-run-default-action (citar--get-entry "stanley2021"))
*** Eval error ***  No associated resources
ELISP> (citar-run-default-action (list (citar--get-entry "stanley2021")))
*** Eval error ***  Wrong type argument: stringp, ("file" . "/home/bruce/Zotero/storage/7HH79WKH/2021 - Movie at the Ellipse A Study in Fascist Propagand.html")
```

**Comment by jkroes** [comment]
Oddly, I was experiencing this issue even in vanilla Emacs. But after fixing a seemingly unrelated coding system issue, the issue disappeared. Sorry for the false positive. Doom Emacs is a pain sometime.

**Comment by jkroes** [comment]
The backtrace was something I generated. The error message I get is the same error message on the last line of your example above: "Wrong type argument: stringp..."

The issue is `citar-run-default-action`. It expects its argument to be a list, but `embark-act` doesn't pass it a list. It passes the key as a string.

**Comment by bdarcus** [comment]
> The issue is `citar-run-default-action`. It expects its argument to be a list, but `embark-act` doesn't pass it a list. It passes the key as a string.

So the solution is to modify that function to check the type, and if a string, then wrap it in a list?

Odd this never came up before; wonder if there was a change in embark?

**Comment by bdarcus** [comment]
Wait a minute; perhaps this issue is related: #583?

**Comment by jkroes** [comment]
> > The issue is `citar-run-default-action`. It expects its argument to be a list, but `embark-act` doesn't pass it a list. It passes the key as a string.
> 
> So the solution is to modify that function to check the type, and if a string, then wrap it in a list?
> 
> Odd this never came up before; wonder if there was a change in embark?

That's right. This is exactly what `embark-dwim` does before it calls `citar-run-default-action`.

**Comment by bdarcus** [comment]
Right; but this code has been there for months, and I'm pretty sure was working. So I'm just confused.

In general, the embark integration stuff in this area confuses me, and seems easy to break.

**Comment by localauthor** [comment]
> I'm admittedly a bit tired, so may just be missing something obvious, but ...
> 
> So is the issue `embark-act`, or is it `citar-run-default-action`?
> 
> I don't have `citar-open` as my default action. If I change it to that, and hit RET after `embark-act`, I do get odd behavior, but not any error or backtrace as you do.
> 
> And then there's this, which I can't understand (this resource does have an associated file, which is presented when running `citar-open` directly, but not, as you can see, `citar-run-default-action`).
> 
> ```emacs-lisp
> ELISP> (citar--get-entry "stanley2021")
> (("file" . "/home/bruce/Zotero/storage/7HH79WKH/2021 - Movie at the Ellipse A Study in Fascist Propagand.html")
>  ("keywords" . "2021-01-06,far right,fascism,insurrection,media,riots,Trump,United States,Washington (DC)")
>  ("url" . "https://www.justsecurity.org/74504/movie-at-the-ellipse-a-study-in-fascist-propaganda/")
>  ("date" . "2021-02-04T13:30:28+00:00")
>  ("author" . "Stanley, Jason")
>  ("title" . "Movie at the {{Ellipse}}: A Study in Fascist Propaganda")
>  ("=type=" . "online")
>  ("=key=" . "stanley2021"))
> 
> ELISP> (citar-run-default-action (citar--get-entry "stanley2021"))
> *** Eval error ***  No associated resources
> ELISP> (citar-run-default-action (list (citar--get-entry "stanley2021")))
> *** Eval error ***  Wrong type argument: stringp, ("file" . "/home/bruce/Zotero/storage/7HH79WKH/2021 - Movie at the Ellipse A Study in Fascist Propagand.html")
> ```
> 
> Do you see the specific problem in the code?

Is the problem here that `citar-run-default-action` takes a `keys-entries` argument, while the above is passing it just an entry (via `citar--get-entry`)

**Comment by bdarcus** [comment]
Yes, I realized subsequently.

**Comment by jkroes** [comment]
> Yes, I realized subsequently.

I actually see the key being passed to `citar-run-default-action`, not an entry. I was confused by the `key-entries` argument, to be honest.

**Comment by oantolin** [comment]
I could not reproduce this one either before or after the fix to embark for #583. But in any case, it's probably worth trying this again with the updated embark to see if the problem persists.

**Comment by bdarcus** [comment]
Since none of us can reproduce this, @jkroes, and I'm going to close this for now.
